File: android/app/src/main/java/app/rentivo/features/home/HomeView.kt
package app.rentivo.features.home

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ListAlt
import androidx.compose.material.icons.automirrored.filled.TrendingUp
import androidx.compose.material.icons.filled.ArrowCircleDown
import androidx.compose.material.icons.filled.ArrowCircleUp
import androidx.compose.material.icons.filled.Apartment
import androidx.compose.material.icons.filled.AutoAwesome
import androidx.compose.material.icons.filled.Bolt
import androidx.compose.material.icons.filled.Build
import androidx.compose.material.icons.filled.CalendarMonth
import androidx.compose.material.icons.filled.Description
import androidx.compose.material.icons.filled.Email
import androidx.compose.material.icons.filled.History
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.Palette
import androidx.compose.material.icons.filled.Percent
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material.icons.filled.Shield
import androidx.compose.material.icons.filled.VpnKey
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.material3.pulltorefresh.PullToRefreshBox
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import app.rentivo.app.AppTab
import app.rentivo.app.LocalAppModel
import app.rentivo.data.DashboardSummary
import app.rentivo.domain.ActivityKind
import app.rentivo.domain.Bill
import app.rentivo.domain.BillID
import app.rentivo.domain.Billing
import app.rentivo.domain.BillingID
import app.rentivo.domain.LoadState
import app.rentivo.domain.Money
import app.rentivo.domain.RecentActivity
import app.rentivo.designsystem.BrandMark
import app.rentivo.designsystem.MoneyText
import app.rentivo.designsystem.PageStateView
import app.rentivo.designsystem.RentivoButton
import app.rentivo.designsystem.RentivoCard
import app.rentivo.designsystem.RentivoColors
import app.rentivo.designsystem.RentivoSpacing
import app.rentivo.designsystem.RentivoTypography
import app.rentivo.designsystem.SectionTitle
import app.rentivo.designsystem.StatusBadge
import app.rentivo.designsystem.capitalizedPTBR
import app.rentivo.designsystem.ptBRCount
import kotlinx.coroutines.launch

/** How many upcoming bills and activity entries the dashboard shows before it stops. */
private const val UpcomingBillsShown = 4
private const val ActivitiesShown = 5

/**
 * The dashboard screen. Port of the `HomeView` body in
 * `ios/Rentivo/Features/Home/HomeView.swift`; its load state is owned by [HomeTab].
 *
 * @param reload re-runs the dashboard load, for pull-to-refresh and the error state's retry.
 * @param onOpenBill pushes a bill's detail screen onto the tab's stack.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun HomeView(
  state: LoadState<HomeData>,
  reload: suspend () -> Unit,
  onOpenBill: (Billing, BillID) -> Unit,
) {
  val scope = rememberCoroutineScope()
  var isRefreshing by remember { mutableStateOf(false) }

  Scaffold(
    containerColor = RentivoColors.paper,
    topBar = {
      TopAppBar(
        title = { Text(text = "Início", style = RentivoTypography.title) },
        actions = { BrandMark(compact = true, modifier = Modifier.padding(end = RentivoSpacing.medium)) },
        colors = TopAppBarDefaults.topAppBarColors(
          containerColor = RentivoColors.paper,
          titleContentColor = RentivoColors.ink,
        ),
      )
    },
  ) { padding ->
    PullToRefreshBox(
      isRefreshing = isRefreshing,
      onRefresh = {
        scope.launch {
          isRefreshing = true
          try {
            reload()
          } finally {
            isRefreshing = false
          }
        }
      },
      modifier = Modifier
        .fillMaxSize()
        .padding(padding),
    ) {
      PageStateView(state = state, retry = { scope.launch { reload() } }) { data ->
        HomeContent(data = data, onOpenBill = onOpenBill)
      }
    }
  }
}

@Composable
private fun HomeContent(
  data: HomeData,
  onOpenBill: (Billing, BillID) -> Unit,
) {
  LazyColumn(
    modifier = Modifier.fillMaxSize(),
    contentPadding = PaddingValues(RentivoSpacing.page),
    verticalArrangement = Arrangement.spacedBy(RentivoSpacing.section),
  ) {
    item { GreetingSection(overdue = data.summary.overdue) }
    item { SummaryGrid(summary = data.summary) }
    if (data.hasBillings) {
      if (data.overdueBills.isNotEmpty()) {
        item { OverdueSection(count = data.overdueBills.size) }
      }
      item { QuickActionsSection() }
      if (data.upcomingBills.isNotEmpty()) {
        item {
          BillsSection(
            title = "Próximas faturas",
            bills = data.upcomingBills.take(UpcomingBillsShown),
            billings = data.billings,
            onOpenBill = onOpenBill,
          )
        }
      }
    } else {
      item { NoBillingsSection() }
    }
    item { ActivitySection(activities = data.activities) }
  }
}

@Composable
private fun GreetingSection(overdue: Money) {
  val app = LocalAppModel.current
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.small)) {
    Text(text = "Olá!", style = RentivoTypography.display, color = RentivoColors.ink)
    Text(
      text = if (app.usesLiveAPI) {
        "Seu portfólio está conectado ao Rentivo."
      } else {
        "Seu portfólio está pronto para a demonstração."
      },
      style = RentivoTypography.body,
      color = RentivoColors.secondaryInk,
    )
    Row(
      modifier = Modifier
        .fillMaxWidth()
        .padding(top = RentivoSpacing.small),
      verticalAlignment = Alignment.CenterVertically,
    ) {
      IconLabel(
        text = "Saldo em atraso",
        icon = Icons.Filled.Schedule,
        modifier = Modifier.weight(1f),
      )
      MoneyText(money = overdue, color = RentivoColors.coral)
    }
  }
}

/**
 * The four dashboard tiles. The iOS `LazyVGrid` of two flexible columns is two rows of equally
 * weighted cards here — a nested `LazyVerticalGrid` cannot be measured inside a `LazyColumn`.
 */
@Composable
private fun SummaryGrid(summary: DashboardSummary) {
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
    Row(horizontalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
      SummaryCard(
        title = "Recebido",
        value = summary.received,
        icon = Icons.Filled.ArrowCircleDown,
        color = RentivoColors.emerald,
        modifier = Modifier.weight(1f),
      )
      SummaryCard(
        title = "Despesas",
        value = summary.expenses,
        icon = Icons.Filled.ArrowCircleUp,
        color = RentivoColors.coral,
        modifier = Modifier.weight(1f),
      )
    }
    Row(horizontalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
      SummaryCard(
        title = "Resultado",
        value = summary.netIncome,
        icon = Icons.AutoMirrored.Filled.TrendingUp,
        color = RentivoColors.blue,
        modifier = Modifier.weight(1f),
      )
      CollectionCard(
        percent = summary.collectionRatePercent,
        modifier = Modifier.weight(1f),
      )
    }
  }
}

@Composable
private fun SummaryCard(
  title: String,
  value: Money,
  icon: ImageVector,
  color: Color,
  modifier: Modifier = Modifier,
) {
  RentivoCard(modifier = modifier) {
    Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.small)) {
      Icon(imageVector = icon, contentDescription = null, tint = color, modifier = Modifier.size(28.dp))
      Text(text = title, style = RentivoTypography.metadata, color = RentivoColors.secondaryInk)
      MoneyText(
        money = value,
        color = RentivoColors.ink,
        // The iOS tile uses `.system(.subheadline, design: .monospaced, weight: .bold)` — smaller
        // than the default money style so four-figure amounts fit two to a row.
        style = RentivoTypography.subheadline.copy(
          fontFamily = FontFamily.Monospace,
          fontWeight = FontWeight.Bold,
        ),
        minimumScaleFactor = 0.7f,
        maxLines = 1,
        contentDescriptionOverride = "$title: ${value.formatted()}",
      )
    }
  }
}

@Composable
private fun CollectionCard(
  percent: Int,
  modifier: Modifier = Modifier,
) {
  RentivoCard(modifier = modifier) {
    Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.small)) {
      Icon(
        imageVector = Icons.Filled.Percent,
        contentDescription = null,
        tint = RentivoColors.lilac,
        modifier = Modifier.size(28.dp),
      )
      Text(
        text = "Taxa de recebimento",
        style = RentivoTypography.metadata,
        color = RentivoColors.secondaryInk,
      )
      Text(text = "$percent%", style = RentivoTypography.money, color = RentivoColors.ink)
    }
  }
}

@Composable
private fun OverdueSection(count: Int) {
  val app = LocalAppModel.current
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
    SectionTitle(title = "Atenção necessária", icon = Icons.Filled.Warning)
    RentivoCard {
      Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
        Text(
          text = "Há ${
            ptBRCount(
              count = count,
              singular = "fatura em acompanhamento",
              plural = "faturas em acompanhamento",
            )
          }",
          style = RentivoTypography.cardTitle,
          color = RentivoColors.ink,
        )
        Text(
          text = "Abra Cobranças para registrar o pagamento ou cancelar a fatura.",
          style = RentivoTypography.subheadline,
          color = RentivoColors.secondaryInk,
        )
        RentivoButton(text = "Ver cobranças", onClick = { app.selectedTab = AppTab.BILLINGS })
      }
    }
  }
}

@Composable
private fun QuickActionsSection() {
  val app = LocalAppModel.current
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
    SectionTitle(title = "Ações rápidas", icon = Icons.Filled.Bolt)
    // This action only switches to the Billings tab — it does not open a create flow (that would
    // require the Billings tab to observe a cross-tab "present create sheet" signal, which lives
    // outside the files this screen owns). Naming it "Ver cobranças" keeps the label honest about
    // what actually happens.
    RentivoButton(onClick = { app.selectedTab = AppTab.BILLINGS }) {
      Icon(imageVector = Icons.AutoMirrored.Filled.ListAlt, contentDescription = null, tint = Color.White)
      Spacer(modifier = Modifier.width(RentivoSpacing.small))
      Text(text = "Ver cobranças", style = RentivoTypography.cardTitle, color = Color.White)
    }
  }
}

@Composable
private fun NoBillingsSection() {
  val app = LocalAppModel.current
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
    SectionTitle(title = "Comece por aqui", icon = Icons.Filled.AutoAwesome)
    RentivoCard {
      Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
        Text(
          text = "Nenhuma cobrança cadastrada ainda",
          style = RentivoTypography.cardTitle,
          color = RentivoColors.ink,
        )
        Text(
          text = "Crie sua primeira cobrança recorrente na aba Cobranças para começar a " +
            "acompanhar recebimentos, despesas e faturas por aqui.",
          style = RentivoTypography.subheadline,
          color = RentivoColors.secondaryInk,
        )
        RentivoButton(text = "Ver cobranças", onClick = { app.selectedTab = AppTab.BILLINGS })
      }
    }
  }
}

@Composable
private fun BillsSection(
  title: String,
  bills: List<Bill>,
  billings: Map<BillingID, Billing>,
  onOpenBill: (Billing, BillID) -> Unit,
) {
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
    SectionTitle(title = title, icon = Icons.Filled.CalendarMonth)
    bills.forEach { bill ->
      val billing = billings[bill.billingID]
      RentivoCard(
        modifier = Modifier
          .testTag("home.bill.card.${bill.id.rawValue}")
          // A bill always arrives with the billing it was listed from; without one there is
          // nothing to open a detail screen against, so the card stays inert.
          .clickable(enabled = billing != null) { billing?.let { onOpenBill(it, bill.id) } },
      ) {
        Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.small)) {
          Row(verticalAlignment = Alignment.Top) {
            Column(
              modifier = Modifier.weight(1f),
              verticalArrangement = Arrangement.spacedBy(RentivoSpacing.tiny),
            ) {
              Text(
                text = billing?.name ?: "Cobrança",
                style = RentivoTypography.cardTitle,
                color = RentivoColors.ink,
              )
              Text(
                text = bill.referenceMonth.label.capitalizedPTBR(),
                style = RentivoTypography.subheadline,
                color = RentivoColors.secondaryInk,
              )
            }
            StatusBadge(status = bill.status)
          }
          Row(verticalAlignment = Alignment.CenterVertically) {
            val dueDate = bill.dueDate
            if (dueDate != null) {
              IconLabel(
                text = "Vence em ${dueDate.displayFormatted}",
                icon = Icons.Filled.CalendarMonth,
                style = RentivoTypography.caption,
              )
            }
            Spacer(modifier = Modifier.weight(1f))
            MoneyText(money = bill.effectiveTotal)
          }
        }
      }
    }
  }
}

@Composable
private fun ActivitySection(activities: List<RecentActivity>) {
  val app = LocalAppModel.current
  Column(verticalArrangement = Arrangement.spacedBy(RentivoSpacing.medium)) {
    SectionTitle(title = "Atividade recente", icon = Icons.Filled.History)
    if (activities.isEmpty()) {
      Text(
        text = if (app.usesLiveAPI) {
          "Nenhuma atividade recente."
        } else {
          "As mudanças feitas na demonstração aparecerão aqui."
        },
        style = RentivoTypography.body,
        color = RentivoColors.secondaryInk,
      )
    } else {
      activities.take(ActivitiesShown).forEach { activity ->
        Row(
          modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = RentivoSpacing.tiny),
          horizontalArrangement = Arrangement.spacedBy(RentivoSpacing.medium),
          verticalAlignment = Alignment.Top,
        ) {
          Icon(
            imageVector = activity.kind.icon,
            contentDescription = null,
            tint = RentivoColors.emerald,
            modifier = Modifier.width(24.dp),
          )
          Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(RentivoSpacing.tiny),
          ) {
            Text(
              text = activity.title,
              style = RentivoTypography.subheadlineEmphasized,
              color = RentivoColors.ink,
            )
            Text(
              text = activity.detail,
              style = RentivoTypography.caption,
              color = RentivoColors.secondaryInk,
            )
          }
        }
      }
    }
  }
}

/** The SwiftUI `Label` pairing used inline (outside section headers): a small icon and its text. */
@Composable
private fun IconLabel(
  text: String,
  icon: ImageVector,
  modifier: Modifier = Modifier,
  style: TextStyle = RentivoTypography.metadata,
) {
  Row(
    modifier = modifier,
    horizontalArrangement = Arrangement.spacedBy(RentivoSpacing.tiny),
    verticalAlignment = Alignment.CenterVertically,
  ) {
    Icon(
      imageVector = icon,
      contentDescription = null,
      tint = RentivoColors.ink,
      modifier = Modifier.size(16.dp),
    )
    Text(text = text, style = style, color = RentivoColors.ink)
  }
}

/** The icon each activity kind renders with, mirroring the iOS SF Symbol per kind. */
private val ActivityKind.icon: ImageVector
  get() = when (this) {
    ActivityKind.BILLING -> Icons.Filled.Home
    ActivityKind.BILL -> Icons.Filled.Description
    ActivityKind.EXPENSE -> Icons.Filled.Build
    ActivityKind.ORGANIZATION -> Icons.Filled.Apartment
    ActivityKind.INVITATION -> Icons.Filled.Email
    ActivityKind.SECURITY -> Icons.Filled.Shield
    ActivityKind.API_KEY -> Icons.Filled.VpnKey
    ActivityKind.THEME -> Icons.Filled.Palette
  }

File: macos/Rentivo/Features/Home/HomeView.swift
import RentivoCore
import SwiftUI

/// Everything the Início dashboard renders, resolved by a single `load()` so the screen never
/// shows a half-populated mix of fresh and stale sections.
private struct HomeData: Sendable {
  let summary: DashboardSummary
  let overdueBills: [Bill]
  let upcomingBills: [Bill]
  let billingNames: [BillingID: String]
  let activities: [RecentActivity]
  let hasBillings: Bool
}

/// The pure derivations behind the dashboard. iOS keeps these inline in `HomeView.load()` and in
/// the view bodies; on macOS they are lifted out so the bill partitioning, the ordering rule for
/// undated bills, and the copy that switches on live-vs-demo can be asserted without a view.
enum HomeDashboard {
  /// A bill is upcoming while it is still on its way to being paid. `paid`, `cancelled`, and
  /// `delayedPayment` are all excluded: the first two are finished and the third belongs to
  /// "Atenção necessária" instead.
  static let upcomingStatuses: Set<BillStatus> = [.draft, .published, .sent]

  static func overdueBills(in bills: [Bill]) -> [Bill] {
    bills.filter { $0.status == .delayedPayment }
  }

  static func upcomingBills(in bills: [Bill]) -> [Bill] {
    bills.filter { upcomingStatuses.contains($0.status) }.sorted {
      // A bill with no due date has nothing to be "upcoming" against, so it sorts after
      // every dated bill rather than ahead of them.
      switch ($0.dueDate, $1.dueDate) {
      case let (lhs?, rhs?): return lhs < rhs
      case (_?, nil): return true
      case (nil, _?): return false
      case (nil, nil): return false
      }
    }
  }

  static func greetingSubtitle(usesLiveAPI: Bool) -> String {
    usesLiveAPI
      ? "Seu portfólio está conectado ao Rentivo."
      : "Seu portfólio está pronto para a demonstração."
  }

  static func emptyActivityMessage(usesLiveAPI: Bool) -> String {
    usesLiveAPI
      ? "Nenhuma atividade recente."
      : "As mudanças feitas na demonstração aparecerão aqui."
  }
}

struct HomeView: View {
  @Environment(AppModel.self) private var app
  @State private var state: LoadState<HomeData> = .idle
  @State private var refresh = RefreshActivity()

  var body: some View {
    PageStateView(state: state) { data in
      HomeContent(data: data)
    } retry: {
      await load()
    }
    .rentivoPage()
    .navigationTitle("Início")
    .toolbar {
      ToolbarItem(placement: .primaryAction) {
        RefreshToolbarButton(
          activity: refresh,
          help: "Atualizar o painel",
          accessibilityIdentifier: "home.refresh"
        ) {
          await load()
        }
      }
      ToolbarItem(placement: .primaryAction) {
        BrandMark(compact: true)
      }
    }
    .task(id: app.dataRevision) { await load() }
  }

  private func load() async {
    state.prepareForRefresh()
    do {
      // The summary and the portfolio are independent requests, so the dashboard waits for the
      // slower of the two rather than for their sum. `RepositoryBox` is what carries a
      // main-actor repository into the child tasks.
      let dashboard = RepositoryBox(app.dependencies.dashboard)
      let billingsRepository = RepositoryBox(app.dependencies.billings)
      async let summaryRequest = dashboard.repository.dashboardSummary()
      async let billingsRequest = billingsRepository.repository.listBillings()
      let (summary, billings) = try await (summaryRequest, billingsRequest)
      let bills = try await BillLoading.billsByBilling(
        for: billings, using: app.dependencies.bills
      ).flatMap(\.bills)
      let names = Dictionary(uniqueKeysWithValues: billings.map { ($0.id, $0.name) })
      let data = HomeData(
        summary: summary,
        overdueBills: HomeDashboard.overdueBills(in: bills),
        upcomingBills: HomeDashboard.upcomingBills(in: bills),
        billingNames: names,
        activities: app.dependencies.activities.recentActivities,
        hasBillings: !billings.isEmpty
      )
      // The dashboard (summary cards, activity) is always meaningful, even
      // with zero billings — show it with zeroed cards plus an explainer
      // section rather than replacing the whole screen with a generic empty
      // state that has no create action on this screen.
      state = .loaded(data)
    } catch {
      state.settleFailure(error, reportingTo: app)
    }
  }
}

/// Value-based route to a bill's detail screen. The links are value-based (and
/// not `NavigationLink { BillDetailView(...) }`) because acting on a bill from
/// its detail screen — publishing, sending, marking it paid — reloads the
/// dashboard and drops that bill out of "Próximas faturas". A view-based link
/// would be torn down along with its row and pop the user back to Início
/// mid-flow; a pushed value survives its source disappearing.
///
/// Internal rather than `private` (as on iOS) because `BillRouteTests` asserts on the route's
/// equality and hashing directly — the very properties `navigationDestination` relies on.
struct BillRoute: Hashable {
  let billingID: BillingID
  let billID: BillID
}

private struct HomeContent: View {
  @Environment(AppModel.self) private var app
  let data: HomeData

  /// Above this width the detail column fits two columns of cards side by side. Below it the
  /// sections stack in the iOS order instead of being squeezed.
  private static let wideLayoutThreshold: CGFloat = 980
  /// Long-form reading width. Cards stretched across a full-screen 27" display would put the
  /// money on one edge and its label on the other.
  private static let contentMaxWidth: CGFloat = 1100
  private static let activityColumnWidth: CGFloat = 360

  var body: some View {
    GeometryReader { proxy in
      let isWide = proxy.size.width >= Self.wideLayoutThreshold
      ScrollView {
        VStack(alignment: .leading, spacing: RentivoSpacing.section) {
          greeting
            .homeSection(index: 0)
          summaryGrid(isWide: isWide)
            .homeSection(index: 1)
          if isWide {
            wideSections
          } else {
            stackedSections
          }
        }
        .padding(RentivoSpacing.page)
        .frame(maxWidth: Self.contentMaxWidth, alignment: .leading)
        .frame(maxWidth: .infinity)
      }
      .accessibilityIdentifier("home.scroll")
    }
    .navigationDestination(for: BillRoute.self) { route in
      BillDetailView(billingID: route.billingID, billID: route.billID)
    }
  }

  /// The bill-facing sections, in the iOS order. Shared by both layouts so the wide window is
  /// purely a rearrangement — never a different dashboard.
  @ViewBuilder
  private var primarySections: some View {
    if data.hasBillings {
      if !data.overdueBills.isEmpty {
        overdueSection
          .homeSection(index: 2)
      }
      quickActions
        .homeSection(index: 3)
      if !data.upcomingBills.isEmpty {
        billsSection(
          title: "Próximas faturas",
          bills: Array(data.upcomingBills.prefix(4))
        )
        .homeSection(index: 4)
      }
    } else {
      noBillingsSection
        .homeSection(index: 2)
    }
  }

  private var wideSections: some View {
    HStack(alignment: .top, spacing: RentivoSpacing.section) {
      VStack(alignment: .leading, spacing: RentivoSpacing.section) {
        primarySections
      }
      .frame(maxWidth: .infinity, alignment: .leading)
      VStack(alignment: .leading, spacing: RentivoSpacing.section) {
        activitySection
          .homeSection(index: 5)
      }
      .frame(width: Self.activityColumnWidth, alignment: .leading)
    }
  }

  private var stackedSections: some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.section) {
      primarySections
      activitySection
        .homeSection(index: 5)
    }
  }

  /// Three cards across once there is room to keep them legible, two otherwise — which is the
  /// iOS layout. A `.adaptive` grid item is deliberately avoided: it collapses to a single
  /// full-width column in a narrow detail pane.
  private func summaryGrid(isWide: Bool) -> some View {
    LazyVGrid(
      columns: Array(
        repeating: GridItem(.flexible(), spacing: RentivoSpacing.medium),
        count: isWide ? 3 : 2
      ),
      spacing: RentivoSpacing.medium
    ) {
      SummaryCard(
        title: "Recebido",
        value: data.summary.received,
        symbol: "arrow.down.circle.fill",
        color: RentivoColors.emerald
      )
      SummaryCard(
        title: "Despesas",
        value: data.summary.expenses,
        symbol: "arrow.up.circle.fill",
        color: RentivoColors.coral
      )
      SummaryCard(
        title: "Resultado",
        value: data.summary.netIncome,
        symbol: "chart.line.uptrend.xyaxis",
        color: RentivoColors.blue
      )
      CollectionCard(percent: data.summary.collectionRatePercent)
    }
  }

  private var greeting: some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.small) {
      Text("Olá!")
        .font(RentivoTypography.display)
        .foregroundStyle(RentivoColors.ink)
      Text(HomeDashboard.greetingSubtitle(usesLiveAPI: app.usesLiveAPI))
        .foregroundStyle(RentivoColors.secondaryInk)
      HStack {
        Label("Saldo em atraso", systemImage: "clock.badge.exclamationmark")
          .font(RentivoTypography.metadata)
        Spacer()
        MoneyText(money: data.summary.overdue, color: RentivoColors.coral)
          .contentTransition(.numericText())
          .animation(.snappy, value: data.summary.overdue)
      }
      .padding(.top, RentivoSpacing.small)
    }
  }

  private var overdueSection: some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
      SectionTitle(title: "Atenção necessária", symbol: "exclamationmark.triangle.fill")
      RentivoCard {
        VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
          Text(
            "Há \(ptBRCount(data.overdueBills.count, singular: "fatura em acompanhamento", plural: "faturas em acompanhamento"))"
          )
          .font(RentivoTypography.cardTitle)
          Text("Abra Cobranças para registrar o pagamento ou cancelar a fatura.")
            .font(RentivoTypography.body)
            .foregroundStyle(RentivoColors.secondaryInk)
          Button("Ver cobranças") { app.selectedTab = .billings }
            .buttonStyle(.borderedProminent)
        }
      }
    }
  }

  private var quickActions: some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
      SectionTitle(title: "Ações rápidas", symbol: "bolt.fill")
      Button {
        app.selectedTab = .billings
      } label: {
        // This action only switches to the Cobranças section — it does not open a
        // create flow (that would require the Cobranças section to observe a
        // cross-section "present create sheet" signal, which lives outside the
        // files this screen owns). Naming it "Ver cobranças" keeps the label
        // honest about what actually happens.
        Label("Ver cobranças", systemImage: "list.bullet.rectangle.fill")
      }
      .buttonStyle(RentivoButtonStyle())
    }
  }

  private var noBillingsSection: some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
      SectionTitle(title: "Comece por aqui", symbol: "sparkles")
      RentivoCard {
        VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
          Text("Nenhuma cobrança cadastrada ainda")
            .font(RentivoTypography.cardTitle)
          Text(
            "Crie sua primeira cobrança recorrente na aba Cobranças para começar a acompanhar recebimentos, despesas e faturas por aqui."
          )
          .font(RentivoTypography.body)
          .foregroundStyle(RentivoColors.secondaryInk)
          Button("Ver cobranças") { app.selectedTab = .billings }
            .buttonStyle(RentivoButtonStyle())
        }
      }
    }
  }

  private func billsSection(title: String, bills: [Bill]) -> some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
      SectionTitle(title: title, symbol: "calendar")
      ForEach(bills) { bill in
        NavigationLink(value: BillRoute(billingID: bill.billingID, billID: bill.id)) {
          RentivoCard {
            VStack(alignment: .leading, spacing: RentivoSpacing.small) {
              HStack(alignment: .top) {
                VStack(alignment: .leading, spacing: RentivoSpacing.tiny) {
                  Text(data.billingNames[bill.billingID] ?? "Cobrança")
                    .font(RentivoTypography.cardTitle)
                  Text(bill.referenceMonth.label.capitalized)
                    .font(RentivoTypography.body)
                    .foregroundStyle(RentivoColors.secondaryInk)
                }
                Spacer()
                StatusBadge(status: bill.status)
              }
              HStack {
                if let dueDate = bill.dueDate {
                  Label("Vence em \(dueDate.displayFormatted)", systemImage: "calendar")
                    .font(RentivoTypography.caption)
                }
                Spacer()
                MoneyText(money: bill.effectiveTotal)
                  .contentTransition(.numericText())
                  .animation(.snappy, value: bill.effectiveTotal)
              }
            }
          }
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("home.bill.card.\(bill.id.rawValue)")
      }
    }
  }

  private var activitySection: some View {
    VStack(alignment: .leading, spacing: RentivoSpacing.medium) {
      SectionTitle(title: "Atividade recente", symbol: "clock.arrow.circlepath")
      if data.activities.isEmpty {
        Text(HomeDashboard.emptyActivityMessage(usesLiveAPI: app.usesLiveAPI))
          .foregroundStyle(RentivoColors.secondaryInk)
      } else {
        ForEach(data.activities.prefix(5)) { activity in
          HStack(alignment: .top, spacing: RentivoSpacing.medium) {
            Image(systemName: activity.kind.symbol)
              .foregroundStyle(RentivoColors.emerald)
              .frame(width: 24)
            VStack(alignment: .leading, spacing: RentivoSpacing.tiny) {
              Text(activity.title)
                .font(RentivoTypography.bodyStrong)
              Text(activity.detail)
                .font(RentivoTypography.caption)
                .foregroundStyle(RentivoColors.secondaryInk)
            }
            Spacer()
          }
          .padding(.vertical, RentivoSpacing.tiny)
        }
      }
    }
  }
}

/// Sections fade and rise into place in reading order. The stagger is deliberately short — the
/// whole dashboard has settled before a pointer can cross it — and collapses to an instant
/// reveal under Reduce Motion.
private struct HomeSectionAppearance: ViewModifier {
  let index: Int
  @Environment(\.accessibilityReduceMotion) private var reduceMotion
  @State private var hasAppeared = false

  private var isVisible: Bool { hasAppeared || reduceMotion }

  func body(content: Content) -> some View {
    content
      .opacity(isVisible ? 1 : 0)
      .offset(y: isVisible ? 0 : 14)
      .onAppear {
        guard !reduceMotion else { return }
        withAnimation(
          .spring(response: 0.42, dampingFraction: 0.86).delay(Double(index) * 0.06)
        ) {
          hasAppeared = true
        }
      }
  }
}

extension View {
  fileprivate func homeSection(index: Int) -> some View {
    modifier(HomeSectionAppearance(index: index))
  }
}

private struct SummaryCard: View {
  let title: String
  let value: Money
  let symbol: String
  let color: Color

  var body: some View {
    RentivoCard {
      VStack(alignment: .leading, spacing: RentivoSpacing.small) {
        Image(systemName: symbol)
          .font(RentivoTypography.icon)
          .foregroundStyle(color)
        Text(title)
          .font(RentivoTypography.metadata)
          .foregroundStyle(RentivoColors.secondaryInk)
        MoneyText(
          money: value,
          color: RentivoColors.ink,
          font: RentivoTypography.moneyCompact,
          minimumScaleFactor: 0.7,
          lineLimit: 1,
          accessibilityLabelOverride: "\(title): \(value.formatted())"
        )
        .contentTransition(.numericText())
        .animation(.snappy, value: value)
      }
    }
  }
}

private struct CollectionCard: View {
  let percent: Int

  var body: some View {
    RentivoCard {
      VStack(alignment: .leading, spacing: RentivoSpacing.small) {
        Image(systemName: "percent")
          .font(RentivoTypography.icon)
          .foregroundStyle(RentivoColors.lilac)
        Text("Taxa de recebimento")
          .font(RentivoTypography.metadata)
          .foregroundStyle(RentivoColors.secondaryInk)
        Text("\(percent)%")
          .font(RentivoTypography.money)
          .foregroundStyle(RentivoColors.ink)
          .contentTransition(.numericText())
          .animation(.snappy, value: percent)
      }
    }
  }
}

extension ActivityKind {
  fileprivate var symbol: String {
    switch self {
    case .billing: "house.fill"
    case .bill: "doc.text.fill"
    case .expense: "wrench.and.screwdriver.fill"
    case .organization: "building.2.fill"
    case .invitation: "envelope.fill"
    case .security: "lock.shield.fill"
    case .apiKey: "key.fill"
    case .theme: "paintpalette.fill"
    }
  }
}
